### PR TITLE
[DBZ-PGYB][yugabyte/yugabyte-db#24200] Execute snapshot in chunks

### PR DESCRIPTION
**Pull request by vaibhav-yb**
## Problem

For very large tables, the default `SELECT *` query can take a really long time to complete leading to longer time for snapshots.

## Solution

This PR aims to implement snapshotting the table in parallel using an inbuilt method `yb_hash_code` to only run the query for a given hash range. The following 2 configuration properties are introduced with this PR:
1. A new `snapshot.mode` called `parallel` - this will behave exactly like `initial_only` but we will have the ability to launch multiple tasks.
2. `primary.key.hash.columns` - this config takes in a comma separated values of the primary key hash component of the table.

> **Note:** When `snapshot.mode` is set to `parallel`, we will not support providing regex in the property `table.include.list` and the user will need to specify the full name of the table in the property. Additionally, we will only allow one table in the `table.include.list` if `snapshot.mode` is `parallel`.

### CODE REVIEW

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -284,6 +294,104 @@ protected Optional<String> getSnapshotSelect(RelationalSnapshotContext<PostgresP
...
+//            catch (SQLException e) {
+//                notificationService.initialSnapshotNotificationService().notifyCompletedTableWithError(snapshotContext.partition,
+//                        snapshotContext.offset,
+//                        table.id().identifier());
+//                throw new ConnectException("Snapshotting of table " + table.id() + " with query " + queries.get(i) + " failed", e);
+//            }
+//        }
+//    }
+
+    protected List<String> getParallelSnapshotQueries(TableId tableId, List<String> columns) {
```
Comment (vaibhav-yb) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java:375]: Method not required now.

File: debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -634,7 +634,7 @@ protected OptionalLong rowCountForTable(TableId tableId) {
         return OptionalLong.empty();
     }
 
-    private Timer getTableScanLogTimer() {
+    protected Timer getTableScanLogTimer() {
```
Comment (vaibhav-yb) [inline debezium-core/src/main/java/io/debezium/relational/RelationalSnapshotChangeEventSource.java:637]: Revert this change.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 3]
```diff
@@ -983,6 +989,12 @@ public static AutoCreateMode parse(String value, String defaultValue) {
     public static final Field SOURCE_INFO_STRUCT_MAKER = CommonConnectorConfig.SOURCE_INFO_STRUCT_MAKER
             .withDefault(PostgresSourceInfoStructMaker.class.getName());
 
+    public static final Field TASK_ID = Field.create("task.id")
```
Comment (vaibhav-yb) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:992]: This change is only to support another logging related change, isn't actually doing much.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 4]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
     @Override
     public List<Map<String, String>> taskConfigs(int maxTasks) {
+        if (props == null) {
+            return Collections.emptyList();
+        }
+
+        if (props.containsKey("snapshot.mode") && props.get("snapshot.mode").equalsIgnoreCase("parallel")) {
+            LOGGER.info("Initialising parallel snapshot consumption");
+
+            final String tableIncludeList = props.get("table.include.list");
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:72]: Instead of using strings for `snapshot.mode`, `table.include.list` and `parallel`, use the static values from PostgresConnectorConfig. For ex: `PostgresConectorConfig.TABLE_INCLUDE_LIST`

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 5]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+            // Perform basic validations.
+            if (tableIncludeList == null) {
+                throw new DebeziumException("No table provided, provide a table in the table.include.list");
+            } else if (tableIncludeList.contains(",")) {
+                // This might indicate the presence of multiple tables in the include list, we do not want that.
+                throw new DebeziumException("parallel snapshot consumption is only supported with one table at a time");
+            }
+
+            // Publication auto create mode should now be for all tables.
+            if (props.containsKey("publication.autocreate.mode") && props.get("publication.autocreate.mode").equalsIgnoreCase("all_tables")) {
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:82]: Same as above.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 6]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+            final String tableIncludeList = props.get("table.include.list");
+            // Perform basic validations.
+            if (tableIncludeList == null) {
+                throw new DebeziumException("No table provided, provide a table in the table.include.list");
+            } else if (tableIncludeList.contains(",")) {
+                // This might indicate the presence of multiple tables in the include list, we do not want that.
+                throw new DebeziumException("parallel snapshot consumption is only supported with one table at a time");
+            }
+
+            // Publication auto create mode should now be for all tables.
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:81]: nit: Publication auto create mode should not be all_tables.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 7]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+            if (props.containsKey("publication.autocreate.mode") && props.get("publication.autocreate.mode").equalsIgnoreCase("all_tables")) {
+                throw new DebeziumException("Snapshot mode parallel is not supported with publication.autocreate.mode all_tables, " +
+                                            "use publication.autocreate.mode=filtered");
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:84]: Can we move these validations to a separate method `validateSingleTableProvidedForSnapshot()`.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 8]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+            }
+
+            // Add configuration for select override.
+            props.put("snapshot.select.statement.overrides", props.get("table.include.list"));
+
+            return getConfigForParallelSnapshotConsumption(maxTasks);
+        }
+
         // this will always have just one task with the given list of properties
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:93]: We have already handled the null case above, can we remove it from here. Also is the comment still valid or do we need to change it?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 9]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+
+        final long rangeSize = (64 * 1024) / maxTasks;
+
+        for (int i = 0; i < maxTasks; ++i) {
+            Map<String, String> taskProps = new HashMap<>(this.props);
+
+            taskProps.put("task.id", String.valueOf(i));
+
+            long lowerBound = i * rangeSize;
+            long upperBound = lowerBound + rangeSize - 1;
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:108]: This logic seems to be faulty and may lead to data loss. Let us assume maxTasks = 10. The upper bound of last task will be 65529, but the max possible hash range is 65535 (xFFFF).  We will be missing out on the range [65530,65535].

We will have to add logic to handle `i == (maxTasks - 1)` separately.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 10]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+            );
+
+            taskConfigs.add(taskProps);
+        }
+
+        return taskConfigs;
+    }
+
+    protected String getQueryForParallelSnapshotSelect(long lowerBound, long upperBound) {
+        return String.format("SELECT * FROM %s WHERE yb_hash_code(%s) >= %d AND yb_hash_code(%s) <= %s",
```
Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:124]: Why are we using %d for lowerBound and %s for upperBound?

**Review by Sumukh-Phalgaonkar — CHANGES_REQUESTED** [review]

**Comment by Sumukh-Phalgaonkar** [comment]
Also could you add some unit tests for this:

1.  Snapshot of a sufficiently large table (data present in every hash partition) with parallel mode.
2. Negative tests where the deployment would fail due to validations, for ex: multiple tables in include list

**Comment by Sumukh-Phalgaonkar** [comment]
In the run with custom image it was observed that the task-id was not getting printed in the snapshot logs. Can it be added to those logs or will it require larger effort?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 11]
```diff
@@ -61,10 +62,70 @@ public void start(Map<String, String> props) {
...
+        }
+
         // this will always have just one task with the given list of properties
         return props == null ? Collections.emptyList() : Collections.singletonList(new HashMap<>(props));
     }
 
+    protected List<Map<String, String>> getConfigForParallelSnapshotConsumption(int maxTasks) {
+        List<Map<String, String>> taskConfigs = new ArrayList<>();
+
+        final long rangeSize = (64 * 1024) / maxTasks;
```
Comment (siddharth2411) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:100]: Wont we miss the hash range when maxTasks is not a multiple of (64*1024)? 
We want to cover the range 0 - 65535. Suppose we have 3 tasks:
Task-1: hash range = 0 - 21844
Task-2: hash range = 21845 - 43689
Task-3: hash range = 43690 - 65534

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java (hunk 10 above)
Comment (siddharth2411) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java]: We are using %d for lower bound whereas %s for upper bound?

Comment (Sumukh-Phalgaonkar) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java]: What will happen if a user only has range partitioned primary key and the parallel snapshot mode is deployed. What if the primary key is a composite like (id HASH, v1 ASC)? Will this still work or should we add another validation that only hash partitioned columns should be provided in `primary.keys` field?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java (hunk 8 above)
Comment (vaibhav-yb) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:93]: Added one more comment and let's keep it like that for now.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 12]
```diff
@@ -983,6 +989,26 @@ public static AutoCreateMode parse(String value, String defaultValue) {
...
+            .withDisplayName("ID of the connector task")
+            .withType(Type.INT)
+            .withImportance(Importance.LOW)
+            .withDescription("Internal use only");
+
+    public static final Field PRIMARY_KEYS = Field.create("primary.keys")
+            .withDisplayName("Comma separated primary key fields")
+            .withType(Type.STRING)
+            .withImportance(Importance.LOW)
+            .withDescription("A comma separated value having all the primary key components")
```
Comment (siddharth2411) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:1002]: Should we update this description to say HASH components of the primary key?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 13]
```diff
@@ -61,10 +62,81 @@ public void start(Map<String, String> props) {
...
+        if (props.containsKey(PostgresConnectorConfig.SNAPSHOT_MODE.name())
+                && props.get(PostgresConnectorConfig.SNAPSHOT_MODE.name())
+                    .equalsIgnoreCase(PostgresConnectorConfig.SnapshotMode.PARALLEL.getValue())) {
+            LOGGER.info("Initialising parallel snapshot consumption");
+
+            final String tableIncludeList = props.get(PostgresConnectorConfig.TABLE_INCLUDE_LIST.name());
+            // Perform basic validations.
+            validateSingleTableProvidedForParallelSnapshot(tableIncludeList);
+
+            // Publication auto create mode should not be for all tables.
```
Comment (siddharth2411) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:78]: Can also move this into a separate function - `validatePublicationForParallelSnapshot`

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java [hunk 14]
```diff
@@ -61,10 +62,81 @@ public void start(Map<String, String> props) {
...
+        final long upperBoundExclusive = 64 * 1024;
+        final long rangeSize = upperBoundExclusive / maxTasks;
+
+        for (int i = 0; i < maxTasks; ++i) {
+            Map<String, String> taskProps = new HashMap<>(this.props);
+
+            taskProps.put(PostgresConnectorConfig.TASK_ID.name(), String.valueOf(i));
+
+            long lowerBound = i * rangeSize;
+            long upperBound = (i == maxTasks - 1) ? upperBoundExclusive - 1 : (lowerBound + rangeSize - 1);
```
Comment (siddharth2411) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/YugabyteDBConnector.java:118]: Can add a comment explaining this special handling for the last task.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java [hunk 15]
```diff
@@ -1116,6 +1116,46 @@ public void shouldHaveBeforeImageOfUpdatedRow() throws InterruptedException {
         assertThat(updateRecordValue.getStruct(Envelope.FieldName.AFTER).getStruct("aa").getInt32("value")).isEqualTo(404);
     }
```
Comment (siddharth2411) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:1118]: Can you add a test for primary.keys config?

**Review by siddharth2411 — CHANGES_REQUESTED** [review]

**Review by Sumukh-Phalgaonkar — APPROVED** [review]

**Review by siddharth2411 — APPROVED** [review]
